File: leetcode/c++/Search_in_Rotated_Sorted_Array_2.cpp

class Solution {
public:
    bool search(int A[], int n, int target) {
        if(n <= 0) return false;
        
        int l = 0;
        int h = n-1;
        
        while(l <= h){
            int m = l + (h-l)/2;
            if(A[m] == target) return true;
            
            if(A[m] < A[h]){
                if(target >= A[m] && target <= A[h]) l = m+1;
                else h = m-1;
            }else if(A[m] == A[h]){
                h--;
            }else{
                if(target >= A[l] && target <= A[m]) h = m-1;
                else l = m + 1;
            }
        }
        
        return false;
    }
};

File: leetcode/java/SearchinRotatedSortedArray_2.java
class Solution {
    public boolean search(int[] nums, int target) {
        if(nums == null) throw new IllegalArgumentException("x");
        int s = 0;
        int e = nums.length-1;
        while(s <= e) {
            int m = s + (e-s)/2;
            if(target == nums[m]) return true;

            if(nums[m] == nums[e]) {
                e--;
            } else if(nums[m] > nums[e]) {
                if(target == nums[s]) return true;
                if(target < nums[m] && target > nums[s]) e = m-1;
                else s = m+1;
            } else {
                if(target == nums[e]) return true;
                if(target > nums[m] && target < nums[e]) s = m+1;
                else e = m-1;
            }
        }

        return false;
    }
}
